File: Week6/ProblemSet6/sentimental-credit/credit.py
from sys import exit


def main():
    cc_num = int(input("Number:"))
    cc_len = digit_length(cc_num)
    if cc_len in [15, 16, 13]:
        legit_card = luhn_algor(cc_num)
        if (legit_card == True):
            print_cc_type(cc_num, cc_len)
            return 0
    print("INVALID")
    return 1


def print_cc_type(cc_num, cc_len):
    if (cc_len == 15):
        cc_num = cc_num // (10**13)  # 13 because 15 - 2
        if (cc_num == 37) or (cc_num == 34):
            print("AMEX")
    elif (cc_len == 16):
        cc_num = cc_num // (10**14)  # 14 because 16 - 2, and 2 because we want to left 2 numbers
        if cc_num in [51, 52, 53, 54, 55]:
            print("MASTERCARD")

        cc_num = cc_num // 10
        if (cc_num == 4):
            print("VISA")
    elif (cc_len == 13):
        cc_num = cc_num // (10**12)
        if (temp2 == 4):
            print("VISA")


def luhn_algor(cc_num):
    sum = 0
    first_last_digit = 0
    while (cc_num > 0):
        sec_last_digit = ((cc_num // 10) % 10)  # Get the second to the last digit
        first_last_digit = first_last_digit + (cc_num % 10)

        # Skip if 0
        if (sec_last_digit == 0):
            cc_num = cc_num // 100
            continue

        sec_last_digit = sec_last_digit * 2
        # Multiplied number must not exceed with 2 digits
        if (sec_last_digit >= 10):
            first_digit = sec_last_digit % 10
            sec_digit = sec_last_digit // 10
            sec_last_digit = first_digit + sec_digit
        sum = sum + sec_last_digit
        cc_num = cc_num // 100  # Remove last two digits
    sum = (first_last_digit + sum) % 10
    if (sum == 0):
        return True  # Legit
    return False  # Not legit


def digit_length(num):
    count = 0
    while (num > 0):
        num = num // 10
        count += 1
    return count


if __name__ == "__main__":
    main()

File: Week1/Cash/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
// cc means Credit Card
// sec means second as in 2nd
// arr means array
// len means length

int get_len_ofNum(long num)
{
    int digits = 0;
    while (num != 0)
    {
        num = num / 10;  // Removes the last digit
        digits++;
    }
    return digits;
}

int get_first_digits(long cc, int n_digits_toGet)
{
    // Decides how many digits to get
    int digits_toGet = 1;  // The amount of zeros determines how many numbers to get
    for (int i = 1; i <= n_digits_toGet; i++)
    {
        digits_toGet = digits_toGet * 10;
    }

    // Reduce the numbers into the desired length of digits
    long first_digits = cc;
    while (first_digits > digits_toGet)
    {
        first_digits = first_digits / 10;
    }
    return first_digits;
}


int print_cardType(long cc)
{
    int len = get_len_ofNum(cc);
    int first_twoDigits = get_first_digits(cc, 2);

    // Determine if its AMEX
    if (first_twoDigits == 34 || 37 == first_twoDigits)
    {
        if (len == 15)
        {
            printf("AMEX\n");
            return 1;
        }
    }

    // Determine if its MASTERCARD
    // Only 51-55 numbers are accepted
    if (first_twoDigits >= 51 && 55 >= first_twoDigits)
    {
        if (len == 16)
        {
            printf("MASTERCARD\n");
            return 1;
        }
    }

    // Determine if its VISA
    int first_digit = get_first_digits(cc, 1);
    if (first_digit == 4)
    {
        if (len == 13 || 16 == len)
        {
            printf("VISA\n");
            return 1;
        }
    }

    // If all tests fails
    return 0;
}

int luhns_algorithm(long cc)
{
    int len = get_len_ofNum(cc);
    int arr_num[len];
    memset(arr_num, 0, len * sizeof(int));
    // ^-- https://stackoverflow.com/questions/3082914/c-compile-error-variable-sized-object-may-not-be-initialized

    long temp_cc = cc;
    int inc = 0;
    while (temp_cc > 0)
    {
        // Get the 2nd to the last digit
        int sec_lastDigit = (temp_cc % 100) / 10;
        arr_num[inc] = sec_lastDigit * 2;
        inc++;


        // Get the last digit
        int lastDigit = temp_cc % 10;
        arr_num[len - inc] = lastDigit;


        // Remove the last two digits
        temp_cc = temp_cc / 100;
    }


    // Add all digits in the stored array
    int sum = 0;
    for (int i = 0; i < len; i++)
    {
        // If 2 digited number, Split it into single digits
        int digit = arr_num[i];
        if (digit >= 10)
        {
            // Code here, assumes that the number is limited to 2 digits
            int left_digit = arr_num[i] % 10;
            int right_digit = arr_num[i] / 10;
            digit = right_digit + left_digit;
        }

        sum = sum + digit;
    }


    // Check the last digit of the sum if its zero
    int last_digit = sum % 10;
    if (last_digit == 0)
    {
        printf("%i", 1);
        return 1;  // card is legit
    }
    printf("%i", 0);
    return 0;  // Card IS NOT LEGIT
}

void check_card(long cc)
{
    // Identify if its American Express, Mastercard or Visa
    // Check the validity of the card
    if (luhns_algorithm(cc) == 0 || 0 == print_cardType(cc))
    {
        printf("INVALID\n");
    }
}

int main(void)
{
    long cc = get_long("Credit Card: ");
    check_card(cc);
}